src/dmd/dsymbolsem.d

```diff
@@ -336,6 +338,113 @@ private FuncDeclaration buildPostBlit(StructDeclaration sd, Scope* sc)
+   class for the parameter and the function : this(ref paramSTC S p) funcStc;
+*/
+private CtorDeclaration generateCopyCtorDeclaration(StructDeclaration sd, const StorageClass paramStc, const StorageClass funcStc)
+{
+    auto fparams = new Parameters();
+    auto structType = sd.type;
+    fparams.push(new Parameter(paramStc | STC.ref_, structType, Id.p, null, null));
+    ParameterList pList = ParameterList(fparams);
+    auto tf = new TypeFunction(pList, structType, LINK.d, STC.ref_);
+    auto ccd = new CtorDeclaration(sd.loc, Loc.initial, STC.ref_, tf, true);
+    ccd.storage_class |= funcStc;
+    ccd.storage_class |= STC.inference;
+    ccd.generated = true;
+    return ccd;
+}
+
+/* Generates a trivial copy constructor body that simply does memberwise
+ * initialization:
+ *
+ *    this.field1 = rhs.field1;
+ *    this.field2 = rhs.field2;
+ *    ...
+ */
+private Statement generateCopyCtorBody(StructDeclaration sd)
+{
+    Loc loc;
+    Expression e;
+    foreach (v; sd.fields)
+    {
+        auto ec = new AssignExp(loc,
+            new DotVarExp(loc, new ThisExp(loc), v),
+            new DotVarExp(loc, new IdentifierExp(loc, Id.p), v));
+        e = Expression.combine(e, ec);
+        //printf("e.toChars = %s\n", e.toChars());
+    }
+    Statement s1 = new ExpStatement(loc, e);
+    return new CompoundStatement(loc, s1);
+}
+
+/* Generates copy constructors for the fields that define copy constructors.
+   The body of all generated copy constructors is the same and it does
+   memberwise initialization. If any initialization in a particular
+   generated copy constructor is not possible (through implicit conversion
+   or copy construction), the generated copy constructor is marked `@disable`.
+ */
+private CtorDeclaration buildCopyCtor(StructDeclaration sd, Scope* sc)
+{
+    if (sd.postblit)
+        return null;
+
+    Dsymbol s = sd.search(sd.loc, Id.copyCtor);
+    if (s)
+        return s.isCopyCtorDeclaration();
+
+    bool fieldCpCtor;
+    // see if any struct members define a copy constructor
+    foreach (v; sd.fields)
+    {
+        if (v.storage_class & STC.ref_)
+            continue;
+        if (v.overlapped)
+            continue;
+        Type tv = v.type.baseElemOf();
+        if (tv.ty != Tstruct)
+            continue;
+        StructDeclaration sdv = (cast(TypeStruct)tv).sym;
+        if (auto copyCtor = sdv.copyCtor)
+        {
+            fieldCpCtor = true;
+            break;
+        }
+    }
+
+    // if any field defines a copy constructor
+    if (fieldCpCtor)
+    {
+        // generate the body that does memberwise initialization
+        const MOD paramMod = MODFlags.wild;
+        const MOD funcMod = MODFlags.wild;
+        auto ccd = generateCopyCtorDeclaration(sd, ModToStc(paramMod), ModToStc(funcMod));
+        //printf("generating for %s\n", ccd.type.toChars());
+        auto copyCtorBody = generateCopyCtorBody(sd);
+        ccd.fbody = copyCtorBody.syntaxCopy();
```

Comment:
Why does it need to be copied?

Reply:
It doesn't work otherwise in the case of nested structs that define copy constructors.